File: go/list-ops/list_ops.go
// Package listops implements basic list operations.
package listops

// IntList is an abstraction of a list of integers which we can define methods on
type IntList []int
type binaryFunc func(int, int) int
type unaryFunc func(int) int
type predicateFunc func(int) bool

func (s IntList) Foldl(fn binaryFunc, initial int) int {
	result := initial

	for _, v := range s {
		result = fn(result, v)
	}

	return result
}

func (s IntList) Foldr(fn binaryFunc, initial int) int {
	result := initial

	for i := s.Length() - 1; i >= 0; i-- {
		result = fn(s[i], result)
	}

	return result
}

func (s IntList) Filter(fn predicateFunc) IntList {
	result := IntList{}

	for _, v := range s {
		if fn(v) {
			result = append(result, v)
		}
	}

	return result
}

func (s IntList) Length() (length int) {
	for range s {
		length++
	}
	return
}

func (s IntList) Map(fn unaryFunc) IntList {
	result := make(IntList, s.Length())

	for i, v := range s {
		result[i] = fn(v)
	}

	return result
}

func (s IntList) Reverse() IntList {
	result := make(IntList, 0, s.Length())

	for i := s.Length() - 1; i >= 0; i-- {
		result = append(result, s[i])
	}

	return result
}

func (s IntList) Append(lst IntList) IntList {
	sLength := s.Length()
	result := make(IntList, sLength+lst.Length())

	copy(result, s)
	copy(result[sLength:], lst)

	return result
}

func (s IntList) Concat(lists []IntList) IntList {
	result := s

	for _, v := range lists {
		result = result.Append(v)
	}

	return result
}

File: python/list-ops/list_ops.py
from collections.abc import Callable
from typing import TypeVar

E = TypeVar("E")
A = TypeVar("A")
T = TypeVar("T")


def append(list1: list[E], list2: list[E]) -> list[E]:
    return concat([list1, list2])


def concat(lists: list[list[E]]) -> list[E]:
    return [item for sublist in lists for item in sublist]


def filter(predicate: Callable[[E], bool], list: list[E]) -> list[E]:
    return [item for item in list if predicate(item)]


def length(list: list[E]) -> int:
    return sum(1 for _ in list)


def map(function: Callable[[E], T], list: list[E]) -> list[T]:
    return [function(item) for item in list]


def foldl(function: Callable[[A, E], A], list: list[E], initial: A) -> A:
    result = initial
    for item in list:
        result = function(result, item)
    return result


def foldr(function: Callable[[A, E], A], list: list[E], initial: A) -> A:
    return foldl(function, reverse(list), initial)


def reverse(list: list[E]) -> list[E]:
    return foldl(lambda acc, item: append([item], acc), list, [])
